contracts/0.8.9/WithdrawalQueue.sol

```diff
@@ -0,0 +1,678 @@
+            etherToWithdraw = request.cumulativeEther;
+            if (_requestId > 0) {
+                shares -= queue[_requestId - 1].cumulativeShares;
+                etherToWithdraw -= queue[_requestId - 1].cumulativeEther;
+            }
+
+            isFinalized = _requestId < finalizedRequestsCounter;
+            isClaimed = request.claimed;
+        }
+    }
+
+    /**
+     * @notice Finalize requests in [`finalizedRequestsCounter`,`_lastRequestIdToFinalize`] range with `_shareRate`
+     * @dev ether to finalize all the requests should be calculated using `calculateFinalizationParams` and sent with
+     * this call as msg.value
+     * @param _lastRequestIdToFinalize request index in the queue that will be last finalized request in a batch
+     * @param _shareRate share/ETH rate for the protocol with 1e27 decimals
+     */
+    function finalize(
+        uint256 _lastRequestIdToFinalize,
+        uint256 _shareRate
+    ) external payable whenResumed onlyRole(FINALIZE_ROLE) {
+        if (_lastRequestIdToFinalize < finalizedRequestsCounter || _lastRequestIdToFinalize >= queue.length) {
+            revert InvalidFinalizationId();
+        }
+        (uint128 ethToWithdraw, ) = _calculateDiscountedBatch(
+            finalizedRequestsCounter,
+            _lastRequestIdToFinalize,
+            _shareRate
+        );
+
+        if (msg.value < ethToWithdraw) revert NotEnoughEther();
+
+        _updateRateHistory(_shareRate, _lastRequestIdToFinalize);
+
+        lockedEtherAmount += _toUint128(msg.value);
+
+        finalizedRequestsCounter = _lastRequestIdToFinalize + 1;
+    }
+
+    /**
+     * @notice calculates the params to fulfill the next batch of requests in queue
+     * @param _lastIdToFinalize last id in the queue to finalize upon
+     * @param _shareRate share rate to finalize requests with
+     *
+     * @return etherToLock amount of eth required to finalize the batch
+     * @return sharesToBurn amount of shares that should be burned on finalization
+     */
+    function calculateFinalizationParams(
+        uint256 _lastIdToFinalize,
+        uint256 _shareRate
+    ) external view returns (uint128 etherToLock, uint128 sharesToBurn) {
+        return _calculateDiscountedBatch(finalizedRequestsCounter, _lastIdToFinalize, _shareRate);
+    }
+
+    /**
+     * @notice Transfer the right to claim withdrawal to another `_newRecipient`
+     * @dev should be called by the old recepient
+     * @param _requestId id of the request subject to change
+     * @param _newRecipient new recipient address for withdrawal
+     */
+    function changeRecipient(uint256 _requestId, address _newRecipient) external {
+        WithdrawalRequest storage request = queue[_requestId];
+
+        if (request.recipient != msg.sender) revert RecipientExpected(request.recipient, msg.sender);
+        if (request.claimed) revert RequestAlreadyClaimed();
+
+        request.recipient = payable(_newRecipient);
+    }
```

Comment:
we check that `recipient` is not zero address in the `requestWithdrawal` method but not in `changeRecipient`